extensions/amp-live-list/0.1/amp-live-list.js

```diff
@@ -341,7 +374,59 @@ export class AmpLiveList extends AMP.BaseElement {
       liveElement.setAttribute('data-tombstone', '');
       // This will empty out its subtree
       liveElement.textContent = '';
+      count++;
     });
+    return count;
+  }
+
+  /**
+   * Remove items from the live DOM if number of valid live items is over
+   * the `max-items-per-page` limit. `data-tombstone`d items are not considered
+   * live items and are ignored in the count.
+   *
+   * @param {!Element} parent
+   * @return {!Promise}
+   */
+  removeOverflowItems_(parent) {
+    const numOfItemsToDelete = this.curNumOfLiveItems_ - this.maxItemsPerPage_;
+
+    if (numOfItemsToDelete < 1) {
+      return Promise.resolve();
+    }
+    const deleteItemsCandidates = [];
+
+    // Walk through the children from last to first.
+    // Only accumulate the items in this loop. Removing them here
+    // will break the prev reference.
+    for (let child = parent.lastElementChild; child;
+        child = child.previousElementSibling) {
+      if (deleteItemsCandidates.length >= numOfItemsToDelete) {
+        break;
+      }
+      if (!this.isChildTombstone_(child)) {
+        deleteItemsCandidates.push(child);
+      }
+    }
+
+    return this.getVsync().runPromise({
+      measure: state => {
+        state.actualDeleteItems = [];
+        // The moment one of the items is in viewport stop deleting.
+        for (let i = 0; i < state.deleteItemsCandidates.length; i++) {
```

Comment:
Isn't the `state.` wrong here?
I'd recommend not using state at all here. Just make `actualDeleteItems` a closured variable. This pattern was a bad idea and only makes sense in 60fps animations.